File: Samples/Login/C++/Src/ViewModel.cpp
////////////////////////////////////////////////////////////////////////////////////////////////////
// NoesisGUI - http://www.noesisengine.com
// Copyright (c) 2013 Noesis Technologies S.L. All Rights Reserved.
////////////////////////////////////////////////////////////////////////////////////////////////////


#include "ViewModel.h"

#include <NsCore/ReflectionImplement.h>
#include <NsGui/PasswordBox.h>


using namespace Login;
using namespace Noesis;
using namespace NoesisApp;


////////////////////////////////////////////////////////////////////////////////////////////////////
ViewModel::ViewModel()
{
    String::Copy(_accountName, sizeof(_accountName), "");
    String::Copy(_message, sizeof(_message), "");
    _loginCommand.SetExecuteFunc(MakeDelegate(this, &ViewModel::Login));
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void ViewModel::SetAccountName(const char* value)
{
    if (!String::Equals(_accountName, value))
    {
        String::Copy(_accountName, sizeof(_accountName), value);
        OnPropertyChanged("AccountName");
    }
}

////////////////////////////////////////////////////////////////////////////////////////////////////
const char* ViewModel::GetAccountName() const
{
    return _accountName;
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void ViewModel::SetMessage(const char* value)
{
    if (!String::Equals(_message, value))
    {
        String::Copy(_message, sizeof(_message), value);
        OnPropertyChanged("Message");
    }
}

////////////////////////////////////////////////////////////////////////////////////////////////////
const char* ViewModel::GetMessage() const
{
    return _message;
}

////////////////////////////////////////////////////////////////////////////////////////////////////
const DelegateCommand* ViewModel::GetLoginCommand() const
{
    return &_loginCommand;
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void ViewModel::Login(BaseComponent* param)
{
    if (CheckCredentials(static_cast<PasswordBox*>(param)))
    {
        SetMessage("LOGIN SUCCESSFUL");
    }
}

////////////////////////////////////////////////////////////////////////////////////////////////////
bool ViewModel::CheckCredentials(PasswordBox* passwordBox)
{
    const char* password = passwordBox->GetPassword();

    if (String::IsNullOrEmpty(_accountName) && !String::IsNullOrEmpty(password))
    {
        SetMessage("ACCOUNT NAME CANNOT BE EMPTY");
        return false;
    }

    if (!String::Equals(_accountName,"noesis") || !String::Equals(password, "12345"))
    {
        SetMessage("ACCOUNT NAME OR PASSWORD IS INCORRECT");
        return false;
    }

    return true;
}

////////////////////////////////////////////////////////////////////////////////////////////////////
NS_BEGIN_COLD_REGION

NS_IMPLEMENT_REFLECTION(ViewModel)
{
    NsProp("AccountName", &ViewModel::GetAccountName, &ViewModel::SetAccountName);
    NsProp("Message", &ViewModel::GetMessage, &ViewModel::SetMessage);
    NsProp("LoginCommand", &ViewModel::GetLoginCommand);
}

File: Samples/Login/UE4/Content/ViewModel.cs
#if UNITY_5_3_OR_NEWER
#define NOESIS
using Noesis;
using UnityEngine;
#else
using System;
using System.Windows.Controls;
#endif

namespace Login
{
    public class ViewModel : NotifyPropertyChangedBase
    {
        public ViewModel()
        {
            LoginCommand = new DelegateCommand(this.Login);
        }

        public DelegateCommand LoginCommand { get; private set; }

        public string AccountName { get; set; }

        private string _message;
        public string Message
        {
            get { return _message; }
            set
            {
                if (_message != value)
                {
                    _message = value;
                    OnPropertyChanged("Message");
                }
            }
        }

        private void Login(object parameter)
        {
            if (CheckCredentials((PasswordBox)parameter))
            {
                Message = "LOGIN SUCCESSFUL";
            }
        }

        private bool CheckCredentials(PasswordBox passwordBox)
        {
            string password = passwordBox.Password;
            if (string.IsNullOrEmpty(AccountName) && !string.IsNullOrEmpty(password))
            {
                Message = "ACCOUNT NAME CANNOT BE EMPTY";
                return false;
            }

            // Verify login and password
            if (AccountName != "noesis" || password != "12345")
            {
                Message = "ACCOUNT NAME OR PASSWORD IS INCORRECT";
                return false;
            }

            return true;
        }
    }
}
